src/common/utxobased/keymanager/keymanager.ts

```diff
@@ -483,149 +573,206 @@ export function derivationLevelScriptHash(
 }
 
 export function xpubToPubkey(args: XPubToPubkeyArgs): string {
-  const network: BitcoinJSNetwork = bip32NetworkFromCoin({
-    coinString: args.coin,
-    sigType: args.type
-  })
   const coin = getCoinFromString(args.coin)
-  const bip32FromBase58Func = coin.bip32FromBase58Func ?? bip32.fromBase58
-  const node: bip32.BIP32Interface = bip32FromBase58Func(args.xpub, network)
-  return node
-    .derive(args.bip44ChangeIndex)
-    .derive(args.bip44AddressIndex)
-    .publicKey.toString('hex')
+  const coinPrefixes = coin.mainnetConstants
+  const maxLength = getMaxLengthOfCoinPrefixes(coinPrefixes)
+  let pubkey: string | undefined
+  for (let prefixIndex = 0; prefixIndex < maxLength; prefixIndex++) {
+    try {
+      const network: BitcoinJSNetwork = bip32NetworkFromCoin({
+        coinString: args.coin,
+        sigType: args.type,
+        prefixIndex
+      })
+      const bip32FromBase58Func = coin.bip32FromBase58Func ?? bip32.fromBase58
+      const node: bip32.BIP32Interface = bip32FromBase58Func(args.xpub, network)
+      pubkey = node
+        .derive(args.bip44ChangeIndex)
+        .derive(args.bip44AddressIndex)
+        .publicKey.toString('hex')
+    } catch (err) {
+      if (isError(err)) {
+        if (err.message === 'Invalid network version') {
+          continue
+        }
+      }
+      throw err
+    }
+  }
+  if (pubkey == null) {
+    throw Error('unable to derive pubkey from xpub')
+  }
+  return pubkey
 }
 
 export function addressToScriptPubkey(args: AddressToScriptPubkeyArgs): string {
-  const legacy = args.legacy ?? false
-  const network: BitcoinJSNetwork = bip32NetworkFromCoin({
-    coinString: args.coin,
-    legacy: args.legacy
-  })
-  const addressType: AddressTypeEnum = guessAddressTypeFromAddress(
-    args.address,
-    network,
-    args.coin,
-    args.addressType
-  )
-  const coinClass = getCoinFromString(args.coin)
-  const cashAddrPrefixes = getCashAddrPrefixes(coinClass)
-  let payment: bitcoin.payments.PaymentCreator
-  switch (addressType) {
-    case AddressTypeEnum.p2pkh:
-      if (
-        (coinClass.mainnetConstants.cashaddr != null && !legacy) ||
-        (coinClass.legacyConstants?.cashaddr != null && legacy)
-      ) {
-        return scriptHashToScriptPubkey({
-          scriptHash: cashAddressToHash(
-            args.address,
-            cashAddrPrefixes
-          ).scriptHash.toString('hex'),
-          scriptType: ScriptTypeEnum.p2pkh,
-          coin: args.coin
-        })
+  const coin = getCoinFromString(args.coin)
+  const cashAddrPrefixes = getCashAddrPrefixes(coin)
+  const coinPrefixes = coin.mainnetConstants
+  const maxLength = getMaxLengthOfCoinPrefixes(coinPrefixes)
+  let scriptPubkey: string | undefined
+
+  for (let prefixIndex = 0; prefixIndex < maxLength; prefixIndex++) {
```

Comment:
Abstract the loop for all the function as a iterator function which accepts a callback.